lib/unison-sqlite/src/Unison/Sqlite/Connection.hs

```diff
@@ -0,0 +1,481 @@
+module Unison.Sqlite.Connection
+  ( -- * Connection management
+    Connection (..),
+    withConnection,
+    openConnection,
+    closeConnection,
+
+    -- * Executing queries
+
+    -- ** Without results
+
+    -- *** With parameters
+    execute,
+    executeMany,
+
+    -- *** Without parameters
+    execute_,
+
+    -- ** With results
+
+    -- *** With parameters
+    queryList,
+    queryListOne,
+    queryMaybe,
+    queryMaybeOne,
+    queryOne,
+    queryOneOne,
+
+    -- **** With checks
+    queryListCheck,
+    queryListOneCheck,
+    queryMaybeCheck,
+    queryMaybeOneCheck,
+    queryOneCheck,
+    queryOneOneCheck,
+
+    -- *** Without parameters
+    queryList_,
+    queryListOne_,
+    queryMaybe_,
+    queryMaybeOne_,
+    queryOne_,
+    queryOneOne_,
+
+    -- **** With checks
+    queryListCheck_,
+    queryListOneCheck_,
+    queryMaybeCheck_,
+    queryMaybeOneCheck_,
+    queryOneCheck_,
+    queryOneOneCheck_,
+
+    -- * Low-level operations
+    withSavepoint,
+    withStatement,
+
+    -- * Exceptions
+    SqliteException (..),
+    SomeShowTypeable (..),
+    ExpectedAtMostOneRowException (..),
+    ExpectedExactlyOneRowException (..),
+  )
+where
+
+import Control.Concurrent (ThreadId, myThreadId)
+import qualified Database.SQLite.Simple as Sqlite
+import qualified Database.SQLite.Simple.FromField as Sqlite
+import qualified Database.SQLite3.Direct as Sqlite (Database (..))
+import Debug.RecoverRTTI (anythingToString)
+import Unison.Prelude
+import Unison.Sqlite.Sql
+import UnliftIO (MonadUnliftIO)
+import UnliftIO.Exception
+
+-- | A /non-thread safe/ connection to a SQLite database.
+data Connection = Connection
+  { name :: String,
+    file :: FilePath,
+    conn :: Sqlite.Connection
+  }
+
+instance Show Connection where
+  show (Connection name file (Sqlite.Connection (Sqlite.Database conn))) =
+    "Connection " ++ show name ++ " " ++ show file ++ " " ++ show conn
+
+-- | Perform an action with a connection to a SQLite database.
+--
+-- Note: the connection is created with @PRAGMA foreign_keys = ON@ automatically, to work around the fact that SQLite
+-- does not automatically enforce foreign key integrity, because it elected to maintain backwards compatibility with
+-- code that was written before the foreign key integrity feature was implemented.
+withConnection ::
+  MonadUnliftIO m =>
+  -- | Connection name, for debugging.
+  String ->
+  -- | Path to SQLite database file.
+  FilePath ->
+  (Connection -> m a) ->
+  m a
+withConnection name file =
+  bracket (openConnection name file) closeConnection
+
+-- | Open a connection to a SQLite database. Prefer 'withConnection'.
+openConnection ::
+  MonadIO m =>
+  -- | Connection name, for debugging.
+  String ->
+  -- | Path to SQLite database file.
+  FilePath ->
+  m Connection
+openConnection name file = do
+  conn0 <- liftIO (Sqlite.open file)
+  let conn = Connection {conn = conn0, file, name}
+  liftIO (execute_ conn "PRAGMA foreign_keys = ON")
```

Comment:
Should we set the journal mode here too?

Reply:
I'm not sure, on one hand it's not something a general-purpose SQLite library would do (unlike setting foreign_keys, which you always want).

But on the other hand, I believe the transaction retry logic in some sense depends on the journal mode, since (for example) a reader does not block a writer in WAL mode. So I'm comfortable with saying "this library only works for WAL-mode databases".

... except we still have that switching to DELETE mode, which we haven't replaced yet. Maybe it's therefore best to leave setting journal mode outside of `openConnection` for now.